Integrate Robot Framework with virtual input

hbatagelo opened the pull request:
With this PR, `test_drag_and_drop.py` now utilizes `WaylandHid`: a Robot Framework library that enables interacting with the virtual pointer through robot keywords.

Saviq commented on mir-ci/mir_ci/test_drag_and_drop.py:
```diff
@@ -1,14 +1,18 @@
+from mir_ci.virtual_pointer import VirtualPointer
 
-APP_PATH = Path(__file__).parent / "clients" / "drag_and_drop_demo.py"
+MIR_CI_PATH = str(Path(__file__).parent)
+APP_PATH = MIR_CI_PATH + "/clients/drag_and_drop_demo.py"
```
Let's use Path :)

```suggestion
MIR_CI_PATH = Path(__file__).parent
APP_PATH = MIR_CI_PATH / "clients/drag_and_drop_demo.py"
```

Saviq commented on mir-ci/mir_ci/test_drag_and_drop.py:
```diff
@@ -1,14 +1,18 @@
+APP_PATH = MIR_CI_PATH + "/clients/drag_and_drop_demo.py"
 STARTUP_TIME = 1.5 * SLOWDOWN
 A_SHORT_TIME = 0.3
+ROBOT_HEADER = f"""*** Settings ***
+Library    {MIR_CI_PATH}/robot/libraries/WaylandHid.py
+"""
```
Let's build a template here:

```suggestion
ROBOT_TEMPLATE = f"""*** Settings ***
Library    {MIR_CI_PATH}/robot/libraries/WaylandHid.py

*** Test Cases ***
{{test_case}}
"""
```

I imagine longer term we'll build it out to a class of its own, wrapping the temporary file handling etc., but this will do for now.

Saviq commented on mir-ci/mir_ci/test_drag_and_drop.py:
```diff
@@ -25,58 +29,72 @@ class TestDragAndDrop:
+        """
+
+        with tempfile.NamedTemporaryFile(mode="w+", suffix=".robot", buffering=1) as robot_file:
+            robot_file.write(ROBOT_HEADER + "\n*** Test Cases ***\n" + robot_test_case)
```
```suggestion
            robot_file.write(ROBOT_TEMPLATE.format(test_case=robot_test_case))
```

Saviq commented on mir-ci/mir_ci/test_drag_and_drop.py:
```diff
@@ -25,58 +29,72 @@ class TestDragAndDrop:
+        """
+
+        with tempfile.NamedTemporaryFile(mode="w+", suffix=".robot", buffering=1) as robot_file:
+            robot_file.write(ROBOT_HEADER + "\n*** Test Cases ***\n" + robot_test_case)
```
```suggestion
            robot_file.write(ROBOT_TEMPLATE.format(test_case=robot_test_case))
```

Saviq reviewed (approved):
Good work :)

I've asked @p-gentili to have a look from their perspective before we land.

p-gentili commented:
Very cool stuff. I guess this hybrid solution makes sense in this case since _pytest_ is doing a lot of handy stuff. Also, I didn't know much about robot's listeners, so TIL.

For the future, if you end up writing _.robot_ files, I'd recommend adding a [robotidy](https://github.com/canonical/zapper/blob/35ecfb636f50cc9f912c67c757673ef12e4ef2d6/zapper-control/tox.ini#L19) check in your tox/pipeline.

hbatagelo commented:
I've changed `Program` to support handling `Path`. Also, for a bit more flexibility in `ROBOT_TEMPLATE`, I added an extra `{library_path}` parameter along with a corresponding `ROBOT_LIBRARY_PATH`.

Saviq commented on mir-ci/mir_ci/robot/libraries/WaylandHid.py:
```diff
@@ -0,0 +1,61 @@
+
+    def _end_test(self, data, result):  # pylint: disable=unused-argument
+        """Listener method called at the end of the test case to disconnect the client."""
+        asyncio.get_event_loop().run_until_complete(self.__aexit__())
```
We should `if self.pointer` this, no? In case we never actually connected?

hbatagelo replied:
True. I'll add the check.

Saviq commented on mir-ci/mir_ci/robot/libraries/WaylandHid.py:
```diff
@@ -0,0 +1,62 @@
+        await self.connect()
+        self.button(Button[button], False)
+
+    async def connect(self):
+        """Connect to the display."""
+        if not self.pointer:
+            await self.__aenter__()  # pylint: disable=unnecessary-dunder-call
+
+    def _end_test(self, data, result):  # pylint: disable=unused-argument
+        """Listener method called at the end of the test case to disconnect the client."""
+        if self.pointer:
+            asyncio.get_event_loop().run_until_complete(self.__aexit__())
```
One last thing... this seems to be going too deep into the bowels of VirtualPointer, how about we add a `.connect` and `.disconnect` there?

Saviq commented on mir-ci/mir_ci/robot/libraries/WaylandHid.py:
```diff
@@ -0,0 +1,62 @@
+    """
+
+    def __init__(self) -> None:
+        display_name = os.environ.get("WAYLAND_DISPLAY")
+        if display_name is None:
+            raise FatalError("WAYLAND_DISPLAY is not defined.")
+        if not display_name:
+            display_name = "wayland-0"
```
Oh and this looks weird, shouldn't we just `.get("WAYLAND_DISPLAY", "wayland-0")`?

hbatagelo replied:
I agree. I misunderstood that we should handle the scenario in which `WAYLAND_DISPLAY` is defined but empty.

hbatagelo commented:
A few additional changes to `WaylandHid`:
- Ensure compatibility with robotframework==6.1.1: a requirement for future integration with the RPA Framework.
- Set the library scope to global. There is no need for connecting/disconnecting in each test case.
